tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/annotations/AnnotationHelper.java

```diff
@@ -0,0 +1,156 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tritium.event.metrics.annotations;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.collect.ImmutableList;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.List;
+
+public final class AnnotationHelper {
+
+    private AnnotationHelper() {
+        throw new UnsupportedOperationException();
+    }
+
+    /**
+     * Annotation as implemented on passed in type or parent of that type, works for both super classes and interfaces.
+     *
+     * @param clazz - Class type to scan for annotations
+     * @param annotation - Annotation type to scan for
+     * @return - First matching annotation found in depth first search, or null if not found
+     */
+    public static <T extends Annotation> T getSuperTypeAnnotation(Class<?> clazz, Class<T> annotation) {
+        if (clazz.isAnnotationPresent(annotation)) {
+            return clazz.getAnnotation(annotation);
+        }
+
+        for (Class<?> ifaces : getParentClasses(clazz)) {
+            T superAnnotation = getSuperTypeAnnotation(ifaces, annotation);
+            if (superAnnotation != null) {
+                return superAnnotation;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Depth first search up the Type hierarchy to find a matching annotation,  Types which do not implement the
+     * specified method signature are ignored.
+     *
+     * @param annotation - Annotation type to scan for
+     * @param clazz - Class type to scan for matching annotations
+     * @param methodSignature - Method to search annotation for
+     * @return - First found matching annotation or null
+     */
+    public static <T extends Annotation> T getMethodAnnotation(
+            Class<T> annotation, Class<?> clazz, MethodSignature methodSignature) {
+
+        Method method;
+        try {
+            method = clazz.getMethod(methodSignature.getMethodName(), methodSignature.getParameterTypes());
+        } catch (NoSuchMethodException e) {
+            return null;
+        }
+
+        if (method.isAnnotationPresent(annotation)) {
+            return method.getAnnotation(annotation);
+        }
+
+        for (Class<?> iface : getParentClasses(clazz)) {
+            T foundAnnotation = getMethodAnnotation(annotation, iface, methodSignature);
+            if (foundAnnotation != null) {
+                return foundAnnotation;
+            }
+        }
+
+        return null;
+    }
+
+    private static List<Class<?>> getParentClasses(Class<?> clazz) {
+        ImmutableList.Builder<Class<?>> builder = ImmutableList.builder();
+        if (clazz.getSuperclass() != null) {
+            builder.add(clazz.getSuperclass());
+        }
+
+        return builder.addAll(Arrays.asList(clazz.getInterfaces())).build();
+    }
```

Comment:
Wondering if we need to traverse the full super-type hierarchy with something like:

```
    @VisibleForTesting
    static Set<Class<?>> getAllParentClasses(Class<?> clazz) {
        ImmutableSet.Builder<Class<?>> builder = ImmutableSet.builder();
        builder.add(clazz.getInterfaces());
        Class<?> superclass = clazz.getSuperclass();
        while (superclass != null) {
            builder.add(superclass.getInterfaces());
            builder.add(superclass);
            superclass = superclass.getSuperclass();
        }
        return builder.build();
    }
```

Reply:
Both times this used it's accessed in recursive methods that climb the type hierarchy.  I'm not sure it's needed and the behavior on on which interface is selected for annotations becomes the parent overwriting the child.